xlators/storage/posix/src/posix-helpers.c

```diff
@@ -1070,6 +1064,21 @@ posix_gfid_set(xlator_t *this, const char *path, loc_t *loc, dict_t *xattr_req,
     else
         ret = posix_handle_soft(this, path, loc, uuid_curr, &stat);
 
+    /* Set the GFID xattr on real_path after link has been populated so
+       that at the time of lookup we can skip gfid (.glusterfs/h1/h2/<file|dir>)
+       validation. If lookup operation has found GFID on real path it
+       means posix_gfid_set set a link with gfid at the time of entry creation
+       the lookup does not need to validate it.
```

Comment:
What happens if the brick dies right after the link operation succeeds but gfid is not set? Subsequent lookup on the file will set a fresh gfid and at the time of unlink it will not delete the earlier file leading to a link leak. This is not a good idea.

Reply:
I don't have clear understanding about the scenario how a lookup will set a fresh gfid.AFAIK the posix has set a gfid based on gfid request from a client.The posix will set a fresh gfid only if it is not receive a correct gfid from the client.Can you please tell me the test case to reproduce the same?.I have tried to reproduce it , in replicate environment the posix
receives correct gfid from the client.

Reply:
check fuse_lookup_resume() and fuse_lookup_cbk() where it does fuse_gfid_set(). Basically if it is a fresh lookup, lookup will set new gfid on the file if one doesn't exist on the file. If it is a revalidate lookup, on failure of lookup a lookup will be sent with a new gfid request to do the same.

Reply:
I went through the code. I think instead of depending on the gfid, depend on the link_count. If stbuf(Not iatt) has more than 1 link, then the file must have gfid with the code without this patch. Even link operation happens with the new path to gfid-path. So this will work in all cases for files. Directories can't be ignored with this approach though.

Reply:
We can not rely on link-count because a user can create hardlink of a file also in that case link-count approach will not work.We can use a below approach to achieve the same.

For specific to avoid a leak of gfid link we can do one thing.
1) In the start of posix_gfid_set we can create a file in a directory <brick_root>/.gfid_link/<gfid_path>
2) After set the xattr on the file/directory successfully remove the file

We can run a background thread to debug this folder. If any gfid is available in the folder job of the
thread is to check and validate the link.If link is stale the thread will remove the gfid link from a backend.

I don't think it will create big drop in entry creation fops and the lookup it will not have any impact. 
Let me know if you are agree with this.

Reply:
@Xavi/@Amar

 Can you please share your view on the same what approach would be good to avoid
 verify link during lookup opetaion or suggest if we can use some other approach?
 
 Pranith has suggested an approach count the hardlink in case of file, for directory
 we can't avoid it.
 
 The other approach i have suggested create a gfid file but it will create
 IOPS.I am not sure how much it would impact performance, we need to calculate it.
 
 I tried to call link validation in background thread also i did not get good performance
 improvement.

Reply:
Basically the issue is that the entry creations in gluster are not crash consistent. Lookup is acting like a recovery mechanism. Without journaling I don't think we can solve it correctly. Even the approach you proposed which creates links in .gfid-link directory is a form of journaling.

Reply:
I have found one another way to avoid gfid_link during lookup without affecting IOPS.After introducing a new xattr at the time of gfid_set we can achieve easily.I have tested the same , i have found very small IOPS difference after introduce a new xattr with similar improvement in lookup operation.

time ./smallfile_cli.py --operation create --threads 16 --file-size 64 --files 1000000 --top /mnt/test1 --files-per-dir 1000000
Without patch 
total threads = 16
total files = 15761000
total data =   961.975 GB
 98.51% of requested files processed, minimum is  90.00
9591.996802 sec elapsed time
1643.140665 files/sec
1643.140665 IOPS
102.696292 MB/sec

real	165m32.633s
user	16m56.291s
sys	24m3.048s

time ./smallfile_cli.py --operation create --threads 16 --file-size 64 --files 1000000 --top /mnt/test1 --files-per-dir 1000000
After populate a new xattr
total threads = 16
total files = 15601700
total data =   952.252 GB
 97.51% of requested files processed, minimum is  90.00
9705.856887 sec elapsed time
1607.452096 files/sec
1607.452096 IOPS
100.465756 MB/sec

real	170m30.785s
user	16m17.504s
sys	23m37.625s

Reply:
I remember some time back we discussed the impact of too many xattrs (more than 512b overall) has on performance (and even suggested compressing them, etc.).

Reply:
Yes, we need to do it.The size of this xattr is just 1 byte.

Reply:
I think the name of the xattr itself also takes up the same space. Unsure.

Reply:
Yes total size will be 14 bytes